Unify analytic solutions interface more

nilsdeppe opened the pull request:
## Proposed changes

- Add `assign_subset(TaggedTuple)` 
- Use a `TaggedTuple<Tags...> variables(x, t, tmpl::list<Tags...>)` method to replace `evolution_variables` and `dt_evolution_variables` in `ScalarWave::PlaneWave` analytic solution. This interface will allow us to overload the `variables` function for retrieving different variables. For example, in EinsteinSolutions we can overload it for the 3+1 variables and the GH variables.
- Adds the `variables` function to Minkowski and KerrSchild einstein solutions
- Use unified interface to reduce the size of `Test_Strahlkorper.cpp` 

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

nilsdeppe committed:
Add subset assignment for Variables from TaggedTuple

nilsdeppe committed:
Add new variables method to ScalarWave::PlaneWave

This method returns a TaggedTuple of the Tensors, which can then be assigned to
a subset of a Variables.

nilsdeppe committed:
KerrSchild provides variables function like Planewave

nilsdeppe committed:
Add variables function to EinsteinSolutions::Minkowski

nilsdeppe committed:
CamelCase tag names inside EinsteinSolutions::KerrSchild

nilsdeppe committed:
Combine Minkowski and KerrSchild tests for Strahlkorper

Because of an interface difference between Minkowski and Strahlkorper several
tests have been duplicated. With the unified interface now in place, this commit
combines the duplicate tests.

kidder commented on tests/Unit/Evolution/DiscontinuousGalerkin/Test_InitializeElement.cpp:
```diff
@@ -41,25 +41,27 @@ struct Var : db::DataBoxTag {
-      const tnsr::I<DataVector, Dim>& x, double t) const noexcept {
-    Variables<tmpl::list<Var>> variables(x.begin()->size());
-    get(get<Var>(variables)) = x.get(0) + t;
+  Variables<tmpl::list<Var>> variables(const tnsr::I<DataVector, Dim>& x,
```
should return TaggedTuple

kidder commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/WaveEquation/Test_PlaneWave.cpp:
```diff
@@ -58,21 +58,39 @@ void check_solution_x(const double kx, const double omega, const DataVector& u,
-                        pw.dpsi_dx(x, t));
-  CHECK_ITERABLE_APPROX(get<ScalarWave::Pi>(pw.evolution_variables(x, t)),
-                        Scalar<DataVector>(-1.0 * pw.dpsi_dt(x, t).get()));
+  CHECK_ITERABLE_APPROX(
```
can you factor out these

kidder commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -254,8 +254,9 @@ class KerrSchild {
   template <typename DataType>
-  tuples::TaggedTupleTypelist<tags<DataType>> solution(
-      const tnsr::I<DataType, 3>& x, double t) const noexcept;
+  tuples::TaggedTupleTypelist<tags<DataType>> variables(
```
do this for all functions

kidder commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Minkowski.hpp:
```diff
@@ -4,7 +4,9 @@
 #pragma once
 
 #include "DataStructures/Tensor/Tensor.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
```
comment why

kidder commented on tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp:
```diff
@@ -315,6 +211,21 @@ SPECTRE_TEST_CASE("Unit.ApparentHorizons.StrahlkorperGr.ExtrinsicCurvature",
                   "[ApparentHorizons][Unit]") {
-  TestRicciScalar::test_schwarzschild();
-  TestRicciScalar::test_minkowski();
+  const double mass = 1.0;
```
also test non-unit mass (check if Geoffrey did this in his #598)

nilsdeppe committed:
Simplify PlaneWave tests

nilsdeppe committed:
Use CAPTURE_PRECISE in CHECK_ITERABLE_APPROX

nilsdeppe committed:
Add GrTags for SqrtDetSpatialMetric and ExtrinsicCurvature

nilsdeppe committed:
Use Tags::dt in Einstein solns & variables() in Minkowski

- Previously there were DtLapse, DtShift, and DtSpatialMetric tags. These are
  replaced with Tags::dt<Lapse>, etc.
- Previously Minkowski had various member functions for getting lapse, shift,
  etc. These are replaced with overloads of the new variables method.

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/WaveEquation/Test_PlaneWave.cpp:
```diff
@@ -32,207 +32,112 @@ inline tnsr::I<double, 3, Frame::Inertial> extract_point_from_coords(
+                    const DataVector& expected_dpsi_dt,
+                    const DataVector& expected_d2psi_dt2,
+                    const DataVector& expected_dpsi_dlast,
+                    const std::array<DataVector, Dim + 1>& expected,
```
`expected_second_derivs` and add comment

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/WaveEquation/Test_PlaneWave.cpp:
```diff
@@ -260,13 +165,31 @@ void test_2d() {
   const auto deserialized_pw = serialize_and_deserialize(pw);
-  check_solution_x(kx, omega, u, deserialized_pw, x, t);
-  check_solution_y(kx, ky, omega, u, deserialized_pw, x, t);
+    check_solution<1>(
```
indentation

nilsdeppe committed:
Fix clang-tidy warning in Domain/Tags.hpp
